경북대 BE 이원준 3단계 - 데이터베이스 적용

lwj0831 opened the pull request.

lwj0831 committed:
docs: README.md 작성

lwj0831 committed:
feat: domain 정의

lwj0831 committed:
chore: domain 변경

Product, Category, ProductImage의 연관관계 구현 추후 ORM도입 시 릭팩토링 예정

lwj0831 committed:
feat: 공통 API응답을 위한 Response 정의

lwj0831 committed:
feat: 상품 단건 조회, 상품 전체 조회 구현

lwj0831 committed:
feat: 상품 등록 api 개발

lwj0831 committed:
feat: 상품 수정 api 개발

lwj0831 committed:
feat: 상품 삭제 api 개발

lwj0831 committed:
feat: GlobalExceptionHandler 정의

lwj0831 committed:
refactor: 상품 조회 api Pagination 추가

lwj0831 committed:
fix: 상품 추가 api 오류 수정

id가 null로 저장되는 bug fix

lwj0831 committed:
refactor: GetProductResDto class->Record로 변경 및 생성자 대신 정적
팩터리 메서드로 변경

-복잡한 로직 없이 단순 dto역할이라 Record로 변경
-메서드 명이 제한적인 생성자 대신 정적 팩터리 메서드로 메서드 역할
분명하게 표현하기 위해 변경

lwj0831 committed:
chore: UpdateProductReqDto에 @NotBlank 추가

lwj0831 committed:
refactor: 예외 발생 부분 Repository->Service layer로 이동

예외 처리는 비즈니스 로직을 담당하는 Service에서 발생시키는 것이 바람직하다고 판단

lwj0831 committed:
refactor: InvalidSortFieldException 추가

lwj0831 committed:
Merge pull request #1 from lwj0831/refactor/product

Refactor/product

lwj0831 committed:
refactor: 상태코드를 위해 ProductController에 ResponseEntity 추가

lwj0831 committed:
refactor(build): thymeleaf 의존성 추가

lwj0831 committed:
refactor(product): GetProductResDto에 price 추가

lwj0831 committed:
refactor: application.properties -> application.yml로 변경 및 thymeleaf 설정 추가

lwj0831 committed:
feat(admin): AdminController내 상품 목록 조회 API개발 및 상품 목록 폼 개발

lwj0831 committed:
fead(admin): 상품 수정 API 및 수정 폼 개발

lwj0831 committed:
feat(admin): 상품 삭제 API 개발

lwj0831 committed:
style: README.md 2단계 내용 추가

lwj0831 committed:
Merge branch 'lwj0831' into lwj0831

lwj0831 committed:
fix: Confilct 해결 시 잘못 입력한 값 수정

lwj0831 committed:
feat(admin): Validation 적용 안되던 문제 해결

lwj0831 committed:
chore(product): @Positive -> @PositiveOrZero로 변경

가격 0이상으로 도메인 조건 변경

lwj0831 committed:
style: Google java style guide 적용

lwj0831 committed:
chore: README.me 불필요한 부분 삭제

lwj0831 committed:
refactor: 불필요한 도메인 클래스 제거

lwj0831 committed:
refactor(product): Product 도메인 클래스 레코드로 변경 및 메서드 리팩토링

- Product 도메인 클래스의 불변성을 보장하기 위해 class에서 Record로 변경
- 기본적인 필드별 검증 로직 추가
- 정적 팩터리 메서드 사용으로 메서드명을 통해 명확한 메서드 동작 파악 및
  검증로직 적용

lwj0831 committed:
fix(product): withId메서드 내 로직 오류 수정

lwj0831 committed:
refactor(exception): ErrorCode 리팩토링

- HttpStatus 필드 추가로 상태코드 정보 포함
- errorCode 필드로 서비스 자체 에러코드 정보 포함
- 의미있는 비즈니스 예외의 명확한 에러메세지 작성

lwj0831 committed:
feat(product): 정렬할 필드값에 따른 전략패턴 도입"

- 기존의 로직은 switch문에 따른 정렬 필드값 분기의 로직이라 복잡도가
  높았음
- 전략 패턴을 도입하여 기존 로직 개선 목적

lwj0831 committed:
refactor: PagedResult 정적 팩터리 메서드 수정

- 생성자로 대체 가능한지 고민해보았으나 this()전 로직 추가 불가능하여
  정적팩터리 메서드로 구현

lwj0831 committed:
chore: ErrorCode 필드명 message->errorMessage로 바뀐 부분 반영

lwj0831 committed:
refactor(product): ProductRepository 추상화 및 InMemoryProductRepository
리팩토링

- 추후 DB 연결 확장성을 고려한 추상화를 위해 ProductRepository
  인터페이스 정의
- InMemoryProductRepository 인터페이스에 맞게 Override 구현
- 메서드 parameter에 대해 Objects.requireNonNull()로 최소한의
  방어코드 구현

lwj0831 committed:
chore(product): NullPointerException 에러 메세지 영어에서 한글로 변경

lwj0831 committed:
chore(product): GetproductResDto 수정(Product 클래스에서 레코드로 변경한 사항 반영)

lwj0831 committed:
refactor(product): ProductService 리팩토링

- CreateProductReqDto, UpdateProdctReqDto에 imageUrl 필드 추가
- Product.of() 정적 팩터리 메서드 대신 DTO를 매개변수로 받는 새로운 정적
  팩터리 메서드 정의하면 좋을지 고민
- 이후 많은 DTO에 해당하는 팩터리 메서드로 인해 Product 클래스가
  무거워질 것이라 생각하여 기존의 of를 사용

lwj0831 committed:
refactor(product): ProductSortStrategyFactory메서드 getStrategy에 예외 발생 명확하게 표현

lwj0831 committed:
refactor(product): 에외 처리 리팩토링

- 기존의 HttpStatus로 대체 가능했던 ErrorCode 변경
- 서비스 내 자체 에러코드 필드 추가
- 비즈니스적으로 의미있는 예외에 대한 HttpStatus, errorCode,
  errorMessage 정의

lwj0831 committed:
refactor(product): ProductContorller 리팩토링

- createProduct시 생성된 객체의 URI 반환하도록 변경(그에 따른 Service,
  Repository도 생성된 Product객체의 Id를 반환하도록 변경)
- 조회, 변경, 삭제 API 개선

lwj0831 committed:
feat: Pagination sort필드 필터링을 위한 MethodArgumentResolver 도입"

- ModelAttribute와 고민하였으나 MethodArgumentResolver로 Pagination
  정렬용 어노테이션 정의하는게 재사용성 측면에서 좋다고 판단
- WebConfig로 SortParamMethodArgumentResolver 추가
- Custom Method Parameter 어노테이션인 SortParam 정의
- dto용 SortInfo 레코드 정의
- SortParamArgumentResolver 커스텀 MethodArugmentResolver 정의

lwj0831 committed:
refactor(product): Pagination 정렬 처리 부분 리팩토링

- Controller에서 커스텀 MethodArgumentResolver 사용하여 정렬 필드 처리하도록 변경
- 전략에 따른 Comparator 선정의 책임 Repository에서 Service로 이동
- ProductSortStrategyFactory내 Comparator 반환하는 정적 팩터리 메서드
  추가

lwj0831 committed:
feat(product): Product validation unit test 작성

lwj0831 committed:
refactor: Pagination 리팩터링

- Spring Data의 Pageable과 유사한 역할의 PageRequest 정의하여
  Pagination관련 파라미터 offset,size, sortInfo 하나의 DTO로 관리하도록 변경
- Controller, Service, Repository에서 관련 로직 변경
- 기존의 sortInfo 전략패턴에 따른 Comparator선정
  Service->Repository단으로 이동
- WebConfig에서 MethodArgumentResolver 이름 변경

lwj0831 committed:
chore(build): lombok 의존성 제거

lwj0831 committed:
refactor: BuisinessException 발생 시 디버깅 용 오류 로깅 출력 기능 추가

lwj0831 committed:
fix(product): 상품 수정 시 id값 null이 되는 bug fix

lwj0831 committed:
test(product): ProductService unit 테스트 코드 작성

- 정상 처리 케이스, 일반적인 오류 케이스 추가
- 동시성 처리 정상 수행 여부 확인할 수 있는 테스트 케이스 추가

lwj0831 committed:
chore(product): 상품 조회 Dto(GetProductResDto)에 누락된 imagUrl필드 추가

lwj0831 committed:
refactor: 관리자 페이지 api 수정 및 form html 수정

-imageUrl 항목 추가

lwj0831 committed:
docs: README.md에 상품 조회 시 Pagingation 지원 부분 기술

lwj0831 committed:
chore: 자잘한 부분 수정

- PageRequest의 pageSize 최소 1이상임을 검증하도록 변경
- GlobalExceptionHandler 로그 출력 시 문구 변경

lwj0831 committed:
chore: favicon.io 요청 처리 빈 값을 보낼 수 있게 추가

lwj0831 committed:
refactor: common을 global 패키지의 하위패키지로 리팩토링

도메인에 종속되지 않는 일반 클래스에 대한 패키지 잘게 분리 위함

lwj0831 committed:
Merge pull request #2 from lwj0831/refactor/product

Refactor/product

lwj0831 committed:
docs: 3단계 구현 기능 목록 README에 추가

lwj0831 committed:
build: H2 의존성 추가 및 Datasource 설정 추가

lwj0831 committed:
feat: DB초기화를 위한 script추가

-Product 테이블 생성
- 초기 테스트 데이터 삽입 쿼리 추가

lwj0831 committed:
feat(product): Jdbc기반  ProductRepository 구현

lwj0831 committed:
config: 프로파일에 따른 Repository 변경을 위한 Configuration 적용

lwj0831 committed:
fix(product): PathVariable 요청값 이름 명시하도록 추가

lwj0831 committed:
test(product): assertAll사용하여 테스트 코드 개선

- 하나의 검증이 실패해도 모든 검증이 다 이루어지도록 변경(여러 검증 실패
  확인 가능)
- 관련된 검증 하나로 그룹화하여 테스트 의도 논리적으로 명확히 표현

lwj0831 committed:
chore(product): errorMessage 한글로 수정

lwj0831 committed:
Merge pull request #3 from lwj0831/refactor/product

Refactor/product

lwj0831 committed:
Merge branch 'lwj0831' into lwj0831

lwj0831 committed:
chore: conflict 해결

lwj0831 committed:
chore: confilct 해결

lwj0831 committed:
ned

liquidjoo commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -51,14 +51,14 @@ public ResponseEntity<Void> createProduct(@Valid @RequestBody CreateProductReqDt
 
   @DeleteMapping("/{id}")
-  public ResponseEntity<Void> deleteProduct(@PathVariable Long id) {
+  public ResponseEntity<Void> deleteProduct(@PathVariable(name="id") Long id) {
```
피드백 반영 좋습니다 👍

liquidjoo commented on README.md:
```diff
@@ -80,3 +83,43 @@ Spring Boot와 Thymeleaf 기반의 상품 관리 웹 애플리케이션입니다
 | POST | `/admin/products/edit/{id}` | 상품 수정 처리 |
 | POST | `/admin/products/delete/{id}` | 상품 삭제 처리 |
 
+# 🧱 3단계 - 데이터베이스 적용
```
요구사항 정리 및 구현 목표 설정 좋습니다 👍

liquidjoo commented on README.md:
```diff
@@ -80,3 +83,43 @@ Spring Boot와 Thymeleaf 기반의 상품 관리 웹 애플리케이션입니다
+- [x] `JdbcTemplate`과 `SimpleJdbcInsert`를 활용한 `ProductRepository` 구현
+- [x] CRUD 기능 완전 구현
+
+### 📌 3. 환경별 Repository 설정 관리
+- [x] **`dev` 프로파일 활성화 시 `JdbcProductRepository` 자동 등록**
+- [x] 테스트 및 운영 환경에서 Repository 구현체 유연한 교체 지원
```
dev 프로파일을 통해서 빈 등록을 했다면 다른 환경도 똑같이 해주는 것이 어떨까요?
기본적으로 InMemory 기반으로 만들어진 것 위주로 돌아가게 될 것 같은데요
이렇게 관리하게 된다면 어떤 일들이 일어날 수 있을까요?

lwj0831 replied:
ProductService에서 ProductRepository를 사용하지 않고 InMemoryProductRepository를 사용하여 Profile에 따른 빈 주입이 일어나지 않고 있었다는 점 확인했습니다. 그 외에 관련 사항들 리팩토링하였습니다!

@Qualifier를 통한 빈 우선순위 방식이 간단할 것이라 생각했지만 이후 여러 구현체가 생긴다면 환경에 따른 복잡한 설정이 어려울 것이라 판단해 Profile을 사용하는 방식으로 구현했습니다

해당 방식으로 환경에 따른 빈 의존이 가능하여 환경 변경 시 별도의 코드 변경할 필요가 없어 핀리하다고 생각했습니다 😊

liquidjoo commented on src/main/java/gift/global/config/ProductRepositoryConfig.java:
```diff
@@ -0,0 +1,19 @@
+  @Profile("dev")
+  public ProductRepository productRepository(DataSource dataSource) {
+    return new JdbcProductRepository(dataSource);
+  }
```
환경에 따른 빈 주입 좋습니다만, 위의 코멘트를 참고해보면 좋을 것 같아요 :)

liquidjoo commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -30,7 +30,7 @@ public ProductController(ProductService productService) {
 
   @GetMapping("/{productId}")
-  public ResponseEntity<GetProductResDto> getProductById(@PathVariable Long productId) {
+  public ResponseEntity<GetProductResDto> getProductById(@PathVariable(name="productId") Long productId) {
```
피드백 반영 좋습니다 👍

liquidjoo commented on src/main/java/gift/product/repository/JdbcProductRepository.java:
```diff
@@ -0,0 +1,104 @@
+
+@Repository
+public class JdbcProductRepository implements ProductRepository{
+  private final NamedParameterJdbcTemplate jdbcTemplate;
```
NamedParameterJdbcTemplate 의 활용 좋습니다!
그렇다면 JdbcTemplate 와 어떤 차이점이 있고 어떤 장점이 존재할까요?

lwj0831 replied:
JdbcTemplate의 경우 바인딩 시 파라미터의 순서까지 정확히 맞춰야한다는 단점이 있는데 NamedParameterJdbcTemplate 의 경우 이름 지정 파라미터 바인딩 방식으로 앞선 방식보다 코드를 잘못 작성하여 오류를 발생시킬 수 있는 위험을 낮출 수 있는 방식이라 NamedParameterJdbcTemplate을 활용했습니다!

liquidjoo commented on src/main/java/gift/product/repository/JdbcProductRepository.java:
```diff
@@ -0,0 +1,104 @@
+  @Override
+  public Long save(Product product) {
+    Objects.requireNonNull(product,"상품은 null이 될 수 없습니다.");
+    SqlParameterSource params = new BeanPropertySqlParameterSource(product);
```
내부 API 활용 좋습니다 👍 
사용함에 있어서 이유를 알고 있으면 더욱 좋을 것 같아요!

lwj0831 replied:
Repository단에서는 커스텀 예외를 던지는 것보단 간단하게 방어적 코드만 추가하는 방식으로도 충분하다고 생각해서 사용했습니다!
비즈니스 검증은 도메인 계층이나 서비스 계층에서 담당하는 것이 좋다고 생각했습니다

liquidjoo commented on src/test/java/gift/product/repository/InMemoryProductRepositoryTest.java:
```diff
@@ -50,11 +53,13 @@ void setUp() {
-    assertThat(foundProduct.get().price()).isEqualTo(10000);
-    assertThat(foundProduct.get().description()).isEqualTo("테스트 상품 설명");
-    assertThat(foundProduct.get().imageUrl()).isEqualTo("http://test.com/image.jpg");
+    assertAll(
```
피드백 반영 좋습니다

liquidjoo reviewed (changes requested):
안녕하세요 원준님 :)
3단계 구현 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다!
확인하고 다시 요청해 주세요 🙇

lwj0831 committed:
fix(product): findAll, update내부 sql 문법 오류 수정 및 RowMapper 변경

- SimpleJdbcInsert를 통해 imageUrl(카멜 스타일)이 image_url(스네이크 스타일)로 저장됨, 이에 따라 아래의 버그 fix
- Product 조회 sql문 내 imageUrl-> image_url로 수정
- 커스텀 RowMapper를 사용하도록 변경하여 mapping field name iamge_url로 변경

lwj0831 committed:
refactor: Profile설정 리팩토링

- ProductService에서 InMemoryProductRepository 사용 ->
  ProductRepository로 변경하여 프로파일에 따른 빈 주입 일어나도록 변경
- ProductRepositoryConfig에서 dev이외에 기본 프로파일 환경인
  default에서의 빈 설정도 추가
- application.yml(공통), application-dev.yml(dev환경용 설정)으로
  프로파일에 따른 application 설정 분리

liquidjoo commented on src/main/java/gift/global/config/ProductRepositoryConfig.java:
```diff
@@ -12,8 +13,13 @@ public class ProductRepositoryConfig {
   @Bean
   @Profile("dev")
-  public ProductRepository productRepository(DataSource dataSource) {
+  public ProductRepository jdbcProductRepository(DataSource dataSource) {
     return new JdbcProductRepository(dataSource);
   }
+  @Bean
+  @Profile("default")
+  public ProductRepository inMemoryProductRepository() {
+    return new InMemoryProductRepository();
+  }
```
피드백 반영 좋습니다 :)
상황별로 달라질 수 있는 점도 고려하면 더욱 좋을 설계 능력을 가질 수 있을 것 같아요!

liquidjoo commented on src/main/java/gift/product/repository/JdbcProductRepository.java:
```diff
@@ -97,8 +96,13 @@ public void deleteById(Long id) {
+        rs.getInt("price"),
+        rs.getString("description"),
+        rs.getString("image_url")
+    );
```
RowMapper 를 별도의 역할로 생각해볼 수 도 있을 것 같네요 :)

liquidjoo reviewed (approved):
안녕하세요 원준님
피드백 반영 잘해주셨어요 👍 
상황에 맞게 잘 고려해서 개발한 점이 인상깊었습니다
저와 함께한 리뷰들이 도움이 되었으면 좋겠습니다
수고하셨습니다 🙇
